File: src/main/kotlin/com/github/dungphan/unityindex/tools/unity/GetProjectContextTool.kt
package com.github.dungphan.unityindex.tools.unity

import com.github.dungphan.unityindex.constants.ToolNames
import com.github.dungphan.unityindex.server.models.ToolCallResult
import com.github.dungphan.unityindex.tools.AbstractMcpTool
import com.github.dungphan.unityindex.tools.schema.SchemaBuilder
import com.intellij.openapi.project.Project
import com.intellij.openapi.vfs.LocalFileSystem
import com.intellij.openapi.vfs.VirtualFile
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonObject

class GetProjectContextTool : AbstractMcpTool() {

    override val requiresPsiSync: Boolean = false

    override val name = ToolNames.GET_PROJECT_CONTEXT

    override val description = """
        Get Unity project context: Unity version, render pipeline, target platforms, company/product name, and installed packages. Essential for any AI agent working with a Unity project — answers "what kind of project is this?" without opening the Editor.

        Parameters: project_path (optional, only needed with multiple projects open).

        Example: {}
    """.trimIndent()

    override val inputSchema: JsonObject = SchemaBuilder.tool()
        .projectPath()
        .build()

    override suspend fun doExecute(project: Project, arguments: JsonObject): ToolCallResult {
        val basePath = project.basePath
            ?: return createErrorResult("Cannot determine project base path")

        val projectDir = LocalFileSystem.getInstance().findFileByPath(basePath)
            ?: return createErrorResult("Project directory not found: $basePath")

        val unityVersion = readUnityVersion(projectDir)
        val projectSettings = readProjectSettings(projectDir)
        val packages = readPackageManifest(projectDir)
        val renderPipeline = detectRenderPipeline(packages)

        return createJsonResult(ProjectContextResult(
            unityVersion = unityVersion,
            renderPipeline = renderPipeline,
            companyName = projectSettings["companyName"],
            productName = projectSettings["productName"],
            targetPlatforms = projectSettings["targetPlatforms"]?.split(",")?.map { it.trim() }?.filter { it.isNotEmpty() } ?: emptyList(),
            scriptingBackend = projectSettings["scriptingBackend"],
            apiCompatibilityLevel = projectSettings["apiCompatibilityLevel"],
            packages = packages,
            projectPath = basePath
        ))
    }

    private fun readUnityVersion(projectDir: VirtualFile): String? {
        val versionFile = projectDir.findChild("ProjectSettings")
            ?.findChild("ProjectVersion.txt") ?: return null
        return try {
            val content = String(versionFile.contentsToByteArray(), Charsets.UTF_8)
            VERSION_REGEX.find(content)?.groupValues?.get(1)
        } catch (_: Exception) {
            null
        }
    }

    private fun readProjectSettings(projectDir: VirtualFile): Map<String, String> {
        val settingsFile = projectDir.findChild("ProjectSettings")
            ?.findChild("ProjectSettings.asset") ?: return emptyMap()
        return try {
            val content = String(settingsFile.contentsToByteArray(), Charsets.UTF_8)
            val result = mutableMapOf<String, String>()

            COMPANY_REGEX.find(content)?.groupValues?.get(1)?.let { result["companyName"] = it }
            PRODUCT_REGEX.find(content)?.groupValues?.get(1)?.let { result["productName"] = it }

            val platforms = mutableListOf<String>()
            for (match in PLATFORM_REGEX.findAll(content)) {
                val platform = match.groupValues[1]
                val enabled = match.groupValues[2]
                if (enabled == "1") platforms.add(platform)
            }
            if (platforms.isNotEmpty()) result["targetPlatforms"] = platforms.joinToString(", ")

            SCRIPTING_BACKEND_REGEX.find(content)?.groupValues?.get(1)?.let { value ->
                result["scriptingBackend"] = if (value == "1") "IL2CPP" else "Mono"
            }

            API_COMPAT_REGEX.find(content)?.groupValues?.get(1)?.let { value ->
                result["apiCompatibilityLevel"] = when (value) {
                    "3" -> ".NET Standard 2.1"
                    "6" -> ".NET Framework"
                    else -> value
                }
            }

            result
        } catch (_: Exception) {
            emptyMap()
        }
    }

    private fun readPackageManifest(projectDir: VirtualFile): List<PackageInfo> {
        val manifestFile = projectDir.findChild("Packages")
            ?.findChild("manifest.json") ?: return emptyList()
        return try {
            val content = String(manifestFile.contentsToByteArray(), Charsets.UTF_8)
            val jsonObj = json.parseToJsonElement(content).let { it as? kotlinx.serialization.json.JsonObject } ?: return emptyList()
            val deps = jsonObj["dependencies"]?.let { it as? kotlinx.serialization.json.JsonObject } ?: return emptyList()

            deps.entries.map { (name, version) ->
                PackageInfo(
                    name = name,
                    version = version.toString().trim('"')
                )
            }.sortedBy { it.name }
        } catch (_: Exception) {
            emptyList()
        }
    }

    private fun detectRenderPipeline(packages: List<PackageInfo>): String {
        val packageNames = packages.map { it.name }.toSet()
        return when {
            "com.unity.render-pipelines.universal" in packageNames -> "URP"
            "com.unity.render-pipelines.high-definition" in packageNames -> "HDRP"
            "com.unity.render-pipelines.core" in packageNames -> "SRP (custom)"
            else -> "Built-in"
        }
    }

    companion object {
        private val VERSION_REGEX = Regex("""m_EditorVersion:\s*(.+)""")
        private val COMPANY_REGEX = Regex("""companyName:\s*(.+)""")
        private val PRODUCT_REGEX = Regex("""productName:\s*(.+)""")
        private val PLATFORM_REGEX = Regex("""enabledNativePlatforms\w*?(\w+):\s*(\d)""")
        private val SCRIPTING_BACKEND_REGEX = Regex("""scriptingBackend:\s*\{[^}]*Standalone:\s*(\d)""")
        private val API_COMPAT_REGEX = Regex("""apiCompatibilityLevelPerPlatform:\s*\{[^}]*Standalone:\s*(\d)""")
    }
}

@Serializable
data class ProjectContextResult(
    val unityVersion: String?,
    val renderPipeline: String,
    val companyName: String?,
    val productName: String?,
    val targetPlatforms: List<String>,
    val scriptingBackend: String?,
    val apiCompatibilityLevel: String?,
    val packages: List<PackageInfo>,
    val projectPath: String
)

@Serializable
data class PackageInfo(
    val name: String,
    val version: String
)

File: vscode-extension/src/tools/unity/getProjectContextTool.ts
import * as fs from "fs";
import * as path from "path";
import { AbstractMcpTool } from "../abstractTool";
import { TOOL_NAMES } from "../../constants";
import { ToolCallResult } from "../../models/jsonRpc";
import { ProjectContext } from "../../server/projectResolver";
import { Args } from "../../utils/args";
import { SchemaBuilder } from "../../utils/schema";

interface PackageInfo {
  name: string;
  version: string;
}

interface ProjectContextResult {
  unityVersion: string | null;
  renderPipeline: string;
  companyName: string | null;
  productName: string | null;
  targetPlatforms: string[];
  scriptingBackend: string | null;
  apiCompatibilityLevel: string | null;
  packages: PackageInfo[];
  projectPath: string;
}

export class GetProjectContextTool extends AbstractMcpTool {
  protected readonly requiresLsp = false;

  readonly name = TOOL_NAMES.GET_PROJECT_CONTEXT;
  readonly description =
    "Get Unity project context: Unity version, render pipeline, target platforms, company/product name, and installed packages. " +
    "Essential for any AI agent working with a Unity project — answers \"what kind of project is this?\" without opening the Editor.";
  readonly inputSchema = SchemaBuilder.tool().projectPath().build();

  protected async doExecute(
    project: ProjectContext,
    _args: Args,
  ): Promise<ToolCallResult> {
    const result: ProjectContextResult = {
      unityVersion: readUnityVersion(project.rootPath),
      renderPipeline: "Built-in",
      companyName: null,
      productName: null,
      targetPlatforms: [],
      scriptingBackend: null,
      apiCompatibilityLevel: null,
      packages: readPackageManifest(project.rootPath),
      projectPath: project.rootPath,
    };
    Object.assign(result, readProjectSettings(project.rootPath));
    result.renderPipeline = detectRenderPipeline(result.packages);
    return this.json(result);
  }
}

const VERSION_RE = /m_EditorVersion:\s*(.+)/;
const COMPANY_RE = /companyName:\s*(.+)/;
const PRODUCT_RE = /productName:\s*(.+)/;
const PLATFORM_RE = /enabledNativePlatforms\w*?(\w+):\s*(\d)/g;
const SCRIPTING_BACKEND_RE = /scriptingBackend:\s*\{[^}]*Standalone:\s*(\d)/;
const API_COMPAT_RE = /apiCompatibilityLevelPerPlatform:\s*\{[^}]*Standalone:\s*(\d)/;

function readUnityVersion(root: string): string | null {
  const f = path.join(root, "ProjectSettings", "ProjectVersion.txt");
  try {
    const content = fs.readFileSync(f, "utf-8");
    const m = VERSION_RE.exec(content);
    return m?.[1].trim() ?? null;
  } catch {
    return null;
  }
}

function readProjectSettings(root: string): Partial<ProjectContextResult> {
  const f = path.join(root, "ProjectSettings", "ProjectSettings.asset");
  try {
    const content = fs.readFileSync(f, "utf-8");
    const result: Partial<ProjectContextResult> = {};
    const company = COMPANY_RE.exec(content);
    if (company) result.companyName = company[1].trim();
    const product = PRODUCT_RE.exec(content);
    if (product) result.productName = product[1].trim();

    const platforms: string[] = [];
    let m: RegExpExecArray | null;
    PLATFORM_RE.lastIndex = 0;
    while ((m = PLATFORM_RE.exec(content)) !== null) {
      if (m[2] === "1") platforms.push(m[1]);
    }
    result.targetPlatforms = platforms;

    const scripting = SCRIPTING_BACKEND_RE.exec(content);
    if (scripting) {
      result.scriptingBackend = scripting[1] === "1" ? "IL2CPP" : "Mono";
    }

    const api = API_COMPAT_RE.exec(content);
    if (api) {
      result.apiCompatibilityLevel =
        api[1] === "3" ? ".NET Standard 2.1" : api[1] === "6" ? ".NET Framework" : api[1];
    }
    return result;
  } catch {
    return {};
  }
}

function readPackageManifest(root: string): PackageInfo[] {
  const f = path.join(root, "Packages", "manifest.json");
  try {
    const content = fs.readFileSync(f, "utf-8");
    const parsed = JSON.parse(content) as { dependencies?: Record<string, string> };
    const deps = parsed.dependencies ?? {};
    return Object.entries(deps)
      .map(([name, version]) => ({ name, version: String(version).replace(/^"|"$/g, "") }))
      .sort((a, b) => a.name.localeCompare(b.name));
  } catch {
    return [];
  }
}

function detectRenderPipeline(packages: PackageInfo[]): string {
  const names = new Set(packages.map((p) => p.name));
  if (names.has("com.unity.render-pipelines.universal")) return "URP";
  if (names.has("com.unity.render-pipelines.high-definition")) return "HDRP";
  if (names.has("com.unity.render-pipelines.core")) return "SRP (custom)";
  return "Built-in";
}
